lib/midcom/helper/imagepopup/handler/upload.php

```diff
@@ -1,161 +1,169 @@
+     */
+    public function _handler_upload($handler_id, array $args, array &$data)
+    {
+        // Ensure we get the correct styles
+        midcom::get()->style->prepend_component_styledir('midcom.helper.imagepopup');
+        midcom::get()->skip_page_style = true;
+		
+		
+        // Get the file
+        reset ($_FILES);
+        $temp = current($_FILES);
+        
+        $accepted_origin = midcom::get()->get_host_name();
+		
+        if(is_uploaded_file($temp['tmp_name']))
+        {
+			         if (isset($_SERVER['HTTP_ORIGIN']))
+			         {
+				            // same-origin requests won't set an origin. If the origin is set, it must be valid.
+				            if (in_array($_SERVER['HTTP_ORIGIN'], $accepted_origin))
+				            {
+				                midcom::get()->header('Access-Control-Allow-Origin: ' . $_SERVER['HTTP_ORIGIN']);
+				            }
+				            else
+				            {
+					               header("HTTP/1.0 403 Origin Denied");
+					               return;
+				            }
+            }
+		
+            if (preg_match("/([^\w\s\d\-_~,;:\[\]\(\).])|([\.]{2,})/", $temp['name']))
+			         {
+				            header("HTTP/1.0 500 Invalid file name.");
+				            return;
+			         }
+		
+			         if (!in_array(strtolower(pathinfo($temp['name'], PATHINFO_EXTENSION)), array("gif", "jpg", "png")))
+			         {
+				            header("HTTP/1.0 500 Invalid extension.");
+				            return;
+			         }
+        }    		
+       
+        // Get the data
+		      $temp_name = $temp['tmp_name'];
+		      $mimetype = $temp['type'];
+				    $parentguid = $args[0];
+		
+		      // Set modified filename 
+		      $filename = $this->get_modify_filename($temp['name']);
+		
+		      // Set the part of the location
+		      $core = "/midcom-serveattachmentguid-";
+		
+		      // Insert the image into database
+		      $this->insert_database($filename, $mimetype, $parentguid);
+		      
+		      // Get target to write the file
+		      $guid = $this->get_data_from_database($filename, $parentguid, "guid");
         
         // Get target to write the file
-        $target = $this->getDataFromDatabase($filename, "target");
+        $target = $this->get_data_from_database($filename, $parentguid, "target");
         
         // Write the file
-        $this->writeTheFile($temp_name, $target);
+        $this->write_the_file($temp_name, $target);
         
         // Make a response for uploadImage() function
         $location = $core. $guid . "/" . $filename;
         
-        // Send a location with uploadImage() function
-		$this->_request_data['location'] =  $location;
-		
-		return true;
-	}
+        // Return image location as JSON
+		      return new midcom_response_json(array('location' => $location));
+    }
 
 	
-	public function _show_upload()
-	{
-		midcom_show_style('midcom_helper_imagepopup_upload');
-	}
+    public function _show_upload()
```

Comment:
this function is unused (because you return `midcom_response_json`), so it can be deleted, and the style element, too